gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/db/FabricDatabase.java

```diff
@@ -1005,6 +1005,162 @@ else if (qVal instanceof AbstractGfxdReplayableMessage) {
+        if (!ds.getName().equals(GfxdConstants.GFXD_DD_DISKSTORE_NAME)) {
+          ds.clearRecoveryMapOfAllOplogs();
+        }
+      }
+    }
+  }
+
+  private void dumpIndexAndRegion(LocalRegion region, DataPolicy dp, GemFireContainer index, LogWriter logger) {
+    ((MemIndex)index.getConglomerate()).dumpIndex("Dumping all indexes");
+    getLocalRegionSz(region, dp, logger, true, false);
+  }
+
+  private void clearAllIndexes(ArrayList<GemFireContainer> uninitializedContainers) {
+    for (GemFireContainer container : uninitializedContainers) {
+      LocalRegion region = container.getRegion();
+      DataPolicy dp = region.getDataPolicy();
+      if (dp == DataPolicy.PERSISTENT_PARTITION || dp == DataPolicy.PERSISTENT_REPLICATE) {
+        GfxdIndexManager gim = (GfxdIndexManager)region.getIndexUpdater();
+        if (gim == null) continue;
+        List<GemFireContainer> allIndexes = gim.getAllIndexes();
+        for (GemFireContainer c : allIndexes) {
+          if (c.isLocalIndex()) {
+            c.getSkipListMap().clear();
+          }
+        }
+      }
+    }
+  }
+
+  private void recreateAllLocalIndexes(final LogWriter logger) {
+    Collection<DiskStoreImpl> diskStores = Misc.getGemFireCache().listDiskStores();
+    for (DiskStoreImpl ds : diskStores) {
+      if (!ds.getName().equals(GfxdConstants.GFXD_DD_DISKSTORE_NAME)) {
+        PersistentOplogSet oplogSet = ds.getPersistentOplogSet(null);
+        ds.resetIndexRecoveryState();
+        // delete all idx file of all oplogs, so second arg as true below
+        ds.scheduleIndexRecovery(oplogSet.getSortedOplogs(), true);
+        logger.info("FabricDatabase: recreateAllLocalIndexes " +
+            "waiting for index re-creation for disk store: " + ds.getName());
+        ds.waitForIndexRecoveryEnd(-1);
+        logger.info("FabricDatabase: recreateAllLocalIndexes " +
+            "index re-creation for disk store: " + ds.getName() + " ended");
+      }
+    }
+  }
+
+  private int getLocalRegionSz(LocalRegion region, DataPolicy dp, final LogWriter logger, boolean dump, boolean throwErrorOnMismatch) {
+    int sz = 0;
+    if (dp == DataPolicy.PERSISTENT_PARTITION) {
+      DiskStoreImpl ds = region.getDiskStore();
+      Collection<AbstractDiskRegion> diskRegions = ds.getAllDiskRegions().values();
+      String regionPath = region.getFullPath();
+      int prId = ((PartitionedRegion)region).getPRId();
+      long regionUUId = region.getRegionUUID();
+      for (AbstractDiskRegion diskReg : diskRegions) {
+        long parentUUid = diskReg.getUUID();
+
+        // check if pr id matches
+        if (parentUUid == regionUUId) {
+          // TODO: Better way to find disk regions of global index's buckets?
+          if (diskReg.getName().contains("____")) continue;
+          if (!dump) {
+            sz += diskReg.getRecoveredEntryCount();
+            int invalidCnt = diskReg.getInvalidOrTombstoneEntryCount();
+            sz -= invalidCnt;
+          } else {
```

Comment:
Yes. That happens in initialize and we want to do all this before this region is declared up in the DS and new entries start coming in the ds.